### DISCUSSION TITLE
Determine path in config for next/image during runtime

### DISCUSSION THREAD

**Discussion by zoltan-improbable** [vercel/next.js#21813] [Help]
When using an external service like imgix to perform image optimization the recommendation is to use relative urls with next/image components and have the following in the config:
```
module.exports = {
  images: {
    loader: 'imgix',
    path: 'https://example.com/myaccount/',
  },
}
```

However I need to switch that path based on my deployment environment (test / staging / production). Using `process.env.IMGIX_URL` for path does not seem to work.
- I cannot rely on NEXT_PUBLIC properties and such since that would mean I need to build different docker images for my deployments per environment and kinda defeats the principles of a ci/cd pipeline.
- I would also like to avoid to use a single S3 bucket with different keys per environment to differentiate.

Is there a way I can use imgix for image optimisation and decide the `path` (baseUrl) at runtime?

**Comment by vicentereig** [comment]
We follow the pattern below along with the  use an `APP_ENV` to denote our multiple environments, for example: `production`, `staging`, `development`, `test`, ...

Somewhere in your deployment pipeline, or even your local terminal, you can fire the build process up with the right config:

```
APP_ENV=staging next build && next export
```
Just call it from your `next.config.js`:

```js
module.exports = {
  // ... 
  images: getImageSettings(),
  // ....
}
```

This function below would give you the `imgix` config for `staging` and `production`, and by default (`test` and `development`) nextjs' loader:

```js
const getImageSettings = () => {
    if ( isProduction() ) {
        return {
            loader: 'imgix', path: 'https://your-production-source.imgix.net/',
        };
    }

    if ( isStaging() ) {
        return {
            loader: 'imgix', path: `https://your-staging-source.imgix.net/`,
        };
    }

    return {};
}
```

`isProduction()` and `isStaging()` are a couple of helper functions to check the value of your `process.env.APP_ENV`. For some of our environments, we do additional checks beyond `process.env.APP_ENV === 'production'` and having a couple of wrapper functions came in handy for testing and keeping the API consistent.
